staging/src/k8s.io/apiserver/pkg/util/version/registry.go

```diff
@@ -0,0 +1,439 @@
+	// returns error if the component is already registered.
+	Register(component string, effectiveVersion MutableEffectiveVersion, featureGate featuregate.MutableVersionedFeatureGate) error
+	// ComponentGlobalsOrRegister would return the registered global variables for the component if it already exists in the registry.
+	// Otherwise, the provided variables would be registered under the component, and the same variables would be returned.
+	ComponentGlobalsOrRegister(component string, effectiveVersion MutableEffectiveVersion, featureGate featuregate.MutableVersionedFeatureGate) (MutableEffectiveVersion, featuregate.MutableVersionedFeatureGate)
+	// AddFlags adds flags of "--emulated-version" and "--feature-gates"
+	AddFlags(fs *pflag.FlagSet)
+	// Set sets the flags for all global variables for all components registered.
+	Set() error
+	// Validate calls the Validate() function for all the global variables for all components registered.
+	Validate() []error
+	// Reset removes all stored ComponentGlobals, configurations, and version mappings.
+	Reset()
+	// SetEmulationVersionMapping sets the mapping from the emulation version of one component
+	// to the emulation version of another component.
+	// Once set, the emulation version of the toComponent will be determined by the emulation version of the fromComponent,
+	// and cannot be set from cmd flags anymore.
+	// For a given component, its emulation version can only depend on one other component, no multiple dependency is allowed.
+	SetEmulationVersionMapping(fromComponent, toComponent string, f VersionMapping) error
+}
+
+type componentGlobalsRegistry struct {
+	componentGlobals map[string]ComponentGlobals
+	mutex            sync.RWMutex
+	// list of component name to emulation version set from the flag.
+	emulationVersionConfig []string
+	// map of component name to the list of feature gates set from the flag.
+	featureGatesConfig map[string][]string
+	// emulationVersionMapping contains the mapping from the emulation version of one component
+	// to the emulation version of another component.
+	emulationVersionMapping map[string]map[string]VersionMapping
+	// componentsWithDependentEmulationVersion stores whether or not a component's EmulationVersion is dependent through mapping on another component.
+	// For a given component, there can only be one mapping from another component.
+	componentsWithDependentEmulationVersion map[string]bool
+	// minCompatibilityVersionMapping contains the mapping from the min compatibility version of one component
+	// to the min compatibility version of another component.
+	minCompatibilityVersionMapping map[string]map[string]VersionMapping
+	// componentsWithDependentMinCompatibilityVersion stores whether or not a component's MinCompatibilityVersion is dependent through mapping on another component
+	// For a given component, there can only be one mapping from another component.
+	componentsWithDependentMinCompatibilityVersion map[string]bool
+}
```

Comment:
while it may structurally work, why the parallel maps instead of a map to a struct keyed by component that  that has the data desired?

Reply:
`componentGlobals map[string]ComponentGlobals` contains the data desired. Other maps are just util maps help setting the final data during setup phase.

Reply:
refactored.